plugins/module_utils/data_merge.py

```diff
@@ -0,0 +1,164 @@
+    def merge_type(self, value):
+        """
+        Set the mean of merging data.
+        Can be one of :
+        - identic : Ensure that the result is identical to the expected data.
+        - present : Ensure that an element in expected data must be present in
+                    the result.  If an element in expected data as different
+                    value or in not present in the base data, it will be set or
+                    added in the result.
+        - absent  : Ensure that an element is absent in the final result. For
+                    the dict, if a key is present in the current dict and in
+                    the expected dict but has different value, the current
+                    value will be kept in the result.
+        """
+        if value not in ['identic', 'present', 'absent']:
+            raise ValueError('`merge_type` can be only one of `identic`, `present` or `absent`')
+        self._merge_type = value
+
+    @property
+    def list_diff_type(self):
+        """
+        Descibe the mean of merging data for a list.
+        """
+        return self._list_diff_type
+
+    @list_diff_type.setter
+    def list_diff_type(self, value):
+        """
+        Descibe the mean of merging data for a list.
+        Can be one of :
+        - value :   Comparing elements in the list by their values. Search if
+                    an element in expect list is present or absent, depending
+                    on the MergType, to the current list.  If it not the case
+                    it will be added of removed. This not check list
+                    recursively, only the first level is checked.
+        - index :   Comparing elements in the list by their index. If
+                    `self.merge_type` is `present`, it will check if the value
+                    of all elements, in the current list as the same value as
+                    the element in the expected list with the same index, and
+                    if it not the case, the value of the element in the
+                    expected list will be set at this index position in the
+                    result list. If `self.merge_type` is `absent`, it will
+                    check if the value of an element in the current list as the
+                    same value as the element with the same index in the
+                    expected list and if it's the case, the element will not be
+                    present in the result.
+        """
+        if value not in ['value', 'index']:
+            raise ValueError('`list_diff_type` can be only one of `value` or `index`')
+        self._list_diff_type = value
+
+    def get_new_merged_data(self, current, expected):
+        # type: (dict | list, dict | list) -> dict | list
+        """
+        Getting the merge of two element if it can't be sure that they have the
+        same type.
+        """
+        if not isinstance(current, (dict, list)) or not isinstance(expected, (dict, list)):
```

Comment:
Instead of `list` and `dict` you should use `Sequence` and `Mapping` (`from ansible.module_utils.common._collections_compat import Mapping, Sequence`).

Reply:
If we do not ensure that the type of element is exactly a `list` or a `dict`, we can get some unexpected behaviour. For example, for `get_new_merged_list`, by replacing `isinstance(current, list) or not isinstance(expected, list)` by `isinstance(current, Sequence) or not isinstance(expected, Sequence)` we can get a `str` as `current` and `expected` which will result to an error.

I think I can integrate `Mapping` or `Sequence` as parameter rather than `list` or `dict` but `DataMerge` will be more complicated and heavy because we will have more case to treat. Is really useful to do that ?

Reply:
> If we do not ensure that the type of element is exactly a `list` or a `dict`, we can get some unexpected behaviour. For example, for `get_new_merged_list`, by replacing `isinstance(current, list) or not isinstance(expected, list)` by `isinstance(current, Sequence) or not isinstance(expected, Sequence)` we can get a `str` as `current` and `expected` which will result to an error.

Well, then make sure it is not `isinstance` of `string_types` (`from ansible.module_utils.six import string_types`).

> I think I can integrate `Mapping` or `Sequence` as parameter rather than `list` or `dict` but `DataMerge` will be more complicated and heavy because we will have more case to treat. Is really useful to do that ?

Since this is used by a plugin, you can easily get data structures in here that are lists or dictionaries, but **not** of type `list` or `dict`. So yes, it is very useful to handle these correctly.